algebird-core/src/main/scala/com/twitter/algebird/ExpHist.scala

```diff
@@ -0,0 +1,342 @@
+   * window of size `windowSize`.
+   */
+  case class Config(epsilon: Double, windowSize: Long) {
+    val k: Int = math.ceil(1 / epsilon).toInt
+    val l: Int = math.ceil(k / 2.0).toInt
+
+    // Returns the last timestamp before the window. any ts <= [the
+    // returned timestamp] is outside the window.
+    def expiration(currTime: Long): Long = currTime - windowSize
+
+    // Drops all buckets with an expired timestamp, based on the
+    // configured window and the supplied current time.
+    def dropExpired(buckets: Vector[Bucket], currTime: Long): (Long, Vector[Bucket]) =
+      ExpHist.dropExpired(buckets, expiration(currTime))
+  }
+
+  /**
+   * Create an empty instance with the supplied Config.
+   */
+  def empty(conf: Config): ExpHist = ExpHist(conf, Vector.empty, 0L, 0L)
+
+  /**
+   *  Generate an instance directly from a number. All buckets in the
+   *  returned ExpHist will have the same timestamp, equal to `ts`.
+   */
+  def from(i: Long, ts: Long, conf: Config): ExpHist = {
+    val buckets = Canonical.bucketsFromLong(i, conf.l).map(Bucket(_, ts))
+    ExpHist(conf, buckets, i, ts)
+  }
+
+  /**
+   * @param buckets [buckets] sorted in DESCENDING order (recent first)
+   * @param cutoff buckets with ts <= cutoff are expired
+   * @return the sum of evicted bucket sizes and the unexpired buckets
+   */
+  def dropExpired(buckets: Vector[Bucket], cutoff: Long): (Long, Vector[Bucket]) = {
+    val (dropped, remaining) = buckets.reverse.span(_.timestamp <= cutoff)
+    (dropped.map(_.size).sum, remaining.reverse)
+  }
+
+  /**
+   * Converts the supplied buckets into a NEW vector of buckets
+   * satisfying this law:
+   *
+   * {{{
+   * rebucket(buckets, desired).map(_.size).sum == desired
+   * }}}
+   *
+   * (rebucket only works if desired.sum == buckets.map(_.size).sum)
+   *
+   * @param buckets [buckets] sorted in DESCENDING order (recent first)
+   * @param desired bucket sizes to rebucket `buckets` into.
+   */
+  private[algebird] def rebucket(buckets: Vector[Bucket], desired: Vector[Long]): Vector[Bucket] =
+    if (desired.isEmpty) Vector.empty
+    else {
+      val input = buckets.dropWhile(_.size == 0)
+      val bucketSize +: tail = desired
+      val remaining = drop(bucketSize, input)
+      input.head.copy(size = bucketSize) +: rebucket(remaining, tail)
+    }
+
+  /**
+   * @param toDrop total count to remove from the left of `input`.
+   * @param input buckets
+   * @return Vector with buckets, or pieces of buckets, with sizes
+   *         totalling `toDrop` items removed from the head. If an
+   *         element wasn't fully consumed, the remainder will be
+   *         stuck back onto the head.
+   */
+  @tailrec private[this] def drop(toDrop: Long, input: Vector[Bucket]): Vector[Bucket] = {
+    val (b @ Bucket(count, _)) +: tail = input
+    (toDrop - count) match {
+      case 0 => tail
+      case x if x < 0 => b.copy(size = -x) +: tail
```

Comment:
this is the only reason I'm not able to use an iterator, this pushing back into the vector. I wonder if we can implement this better with some carry Option. thoughts?